vllm/executor/msgspec_utils.py

```diff
@@ -0,0 +1,36 @@
+from array import array
+from typing import Any, Type
+
+from vllm.sequence import SequenceData
+
+
+def encode_hook(obj: Any) -> Any:
+    """Custom msgspec enc hook that supports array types.
+
+    See https://jcristharif.com/msgspec/api.html#msgspec.msgpack.Encoder
+    """
+    if isinstance(obj, array):
+        return obj.tobytes()
+    if isinstance(obj, SequenceData):
+        # This can be reconstructed from __post_init__.
+        obj._prompt_token_ids_tuple = tuple()
+        obj._cached_all_token_ids = []
+        obj._new_appended_tokens = []
+    else:
+        raise ValueError(f"Unsupported serialization type: {type(obj)}")
+
+
+def decode_hook(type: Type, obj: Any) -> Any:
+    """Custom msgspec dec hook that supports array types.
+
+    See https://jcristharif.com/msgspec/api.html#msgspec.msgpack.Encoder
+    """
+    if type is array:
+        deserialized = array('I')
```

Comment:
Question: do we always return unsigned integer arrays? Why?

Reply:
because values are always unsigned integer, and we don't need to return long (which is 4 bytes).

Reply:
this will double the data size

Reply:
Add the expectation to the docstring?